korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -281,68 +329,313 @@ abstract class BmpSlice(
+
+    fun basePixelPosUnsafe(x: Int, y: Int, out: PointInt = PointInt()): PointInt? = out.also {
+        val offsetX = x - frameOffsetX
+        val offsetY = y - frameOffsetY
+        if (offsetX < 0 || offsetY < 0 || offsetX >= width || offsetY >= height)
+            return null
+
+        it.x = pixelOffsets[0] + pixelOffsets[2] * offsetX + pixelOffsets[3] * offsetY
+        it.y = pixelOffsets[1] + pixelOffsets[4] * offsetY + pixelOffsets[5] * offsetX
+    }
+
     fun readPixels(x: Int, y: Int, width: Int, height: Int, out: RgbaArray = RgbaArray(width * height), offset: Int = 0): RgbaArray {
-        check(x in 0 until this.width)
-        check(y in 0 until this.height)
-        check((x + width) in 0 .. this.width)
-        check((y + height) in 0 .. this.height)
+        check(isValidBasePixelPos(x, y))
+        check(isValidBasePixelPos(x + width - 1, y + height - 1))
         check(out.size >= offset + width * height)
-        bmpBase.readPixelsUnsafe(left + x, top + y, width, height, out, offset)
+        readPixelsUnsafe(x, y, width, height, out, offset)
         return out
     }
 
-    fun getRgba(x: Int, y: Int): RGBA = bmpBase.getRgba(left + x, top + y)
-    fun setRgba(x: Int, y: Int, value: RGBA): Unit = bmpBase.setRgba(left + x, top + y, value)
+    fun readPixelsUnsafe(x: Int, y: Int, width: Int, height: Int, out: RgbaArray, offset: Int = 0) {
+        var n = offset
+        val p = PointInt()
+        for (y0 in 0 until height) for (x0 in 0 until width) {
+            if (basePixelPosUnsafe(x0 + x, y0 + y, p) != null) {
+                out[n++] = bmpBase.getRgba(p.x, p.y)
+            } else {
+                out[n++] = Colors.TRANSPARENT_BLACK
+            }
+        }
+    }
+
+    fun getRgba(x: Int, y: Int): RGBA {
+        basePixelPos(x, y)?.let {
+            return bmpBase.getRgba(it.x, it.y)
+        }
+        return Colors.TRANSPARENT_BLACK
+    }
+
+    fun setRgba(x: Int, y: Int, value: RGBA) {
+        basePixelPos(x, y).also {
+            if (it != null) {
+                bmpBase.setRgba(it.x, it.y, value)
+            } else {
+                if (x < 0 || y < 0 || x >= frameWidth || y >= frameHeight) {
+                    throw IllegalArgumentException("Point $x,$y is not in bounds of slice")
+                }
+                bmpBase = extract()
+                bounds.setBoundsTo(0, 0, bmpBase.width, bmpBase.height)
+                virtFrame?.setBoundsTo(0, 0, bmpBase.width, bmpBase.height)
+                bmpCoords = BmpCoordsWithInstanceBase(
+                    SizeInt(bmpBase.width, bmpBase.height),
+                    0f, 0f, 1f, 0f, 1f, 1f, 0f, 1f
+                )
+                intArrayOf(0, 0, 1, 0, 1, 0).copyInto(pixelOffsets)
+                imageOrientation = ImageOrientation.ORIGINAL
+                bmpBase.setRgba(x, y, value)
+            }
```

Comment:
What's all this code about? If `basePixelPos` is null, shouldn't this just do nothing since it is outside?

Reply:
`null` means, it is outsite the `bmpCoords`. In the else there is a check if it is inside the `virtFrame` and if so the slice will be extracted and the pixel is set in the new extracted Bitmap.

That is the Point
> * if `setRgba()` sets a pixel outside the `BmpCoords` but inside the `virtFrame`, the function extracts a new bitmap and sets it as the new base. To do this, I had to change a few values ​​in properties with `private set()`.